### PR TITLE
fix: Refactor to use constraint names for ON CONFLICT + other fixes

### PR DESCRIPTION
**Pull request by mgagliardo91**
## What?

This PR solves the issue raised at https://github.com/cdaringe/postgraphile-upsert/issues/368. It is broken into two commits. The first one https://github.com/cdaringe/postgraphile-upsert/commit/1440260b74df3b2364554b1555c313c4e87a25c0 is purely updating package dependencies to get it to install - I believe there was some updates recently that caused them to fail on a fresh install.

The second commit https://github.com/cdaringe/postgraphile-upsert/commit/a37e2d1292cd025d786792e54ceccb847ebdbd56 has all of the main logic for the refactor. It may be helpful to view it by commit to reduce some noise.

This PR updates the logic in the resolver to derive which constraint to use for the `ON CONFLICT` statement rather than arbitrarily selecting columns from a flat list of unique constraints. The logic will look at the `where` clause and `input` column names and determine which unique constraint should be included, by name, rather than passing individual column names. The one line change that fixes https://github.com/cdaringe/postgraphile-upsert/issues/368 is [type checking the value](https://github.com/cdaringe/postgraphile-upsert/pull/370/files#diff-e9af2c9157292e93a1b10c99c18c18d5c0c11e2fc6656316f65e2aea9d8ba496R264).

This PR also raised an interesting design decision in this plugin - the where clause conditions were only being used to dictate the `ON CONFLICT` column set, but you could easily pass a different value in the input (or omit it completely) and change the result of the query. 

For example:

```gql
mutation {
  upsertBike(where: {
    weight: 1.1,
    serialKey: "123"
  },
  input: {
    bike: {
      model: "cool-ie deluxe v2"
      weight: 1.1,
      serialKey: "1234"
    }
  }) {
    clientMutationId
  }
}
```

In the above  example, the user dictates a where clause over `weight` and `serialKey`, but the input body of what to insert /update on conflict has a different value for `serialKey` which would yield a new row rather than update the existing. This sort of points to the fact that the `where` clause _could_ really just be a list of columns to upsert on, with the values only being defined in the input to avoid this case.

Instead, this PR adds logic that:

- Adds any column values from the where clause to the upsert logic, if those columns are omitted in the input body
- Throws an error if the value in the input body for a column differs from the where clause condition for the same column

## Future considerations

It may be worth exploring a new format in which the user can pass an upsert like:

```gql
mutation {
  upsertBike(on: [
    WEIGHT,
    SERIAL_KEY,
  ],
  input: {
    bike: {
      model: "cool-ie deluxe v2"
      weight: 1.1,
      serialKey: "1234"
    }
  }) {
    clientMutationId
  }
}
```

which would simplify the usage to only have to indicate the value once, and just specify the columns to apply to the condition, since the where clause values are currently ignored.

### CODE REVIEW

**Comment by cdaringe** [comment]
Thanks for the super clear write up, very helpful. Full agreement on your design considerations. Will review shortly

File: package.json [hunk 1]
```diff
@@ -18,6 +18,7 @@
   "devDependencies": {
     "@commitlint/cli": "17.0.2",
     "@commitlint/config-conventional": "17.0.2",
+    "@timbeyer/commitlint-circle": "^4.1.1",
```
Comment (cdaringe) [inline package.json:21]: why this dep change?

Comment (cdaringe) [inline package.json:21]: why this dep change?

File: package.json [hunk 2]
```diff
@@ -28,12 +29,12 @@
...
     "ava": "4.3.0",
     "bluebird": "3.7.2",
-    "commitlint-circle": "1.0.0",
     "dockerode": "3.3.2",
+    "eslint": "^8.15.0",
     "execa": "5.1.1",
     "freeport": "1.0.5",
     "graphile-build": "4.12.2",
-    "graphql": "16.5.0",
+    "graphql": "15.8.0",
```
Comment (cdaringe) [inline package.json:37]: why is this downgraded?

Comment (cdaringe) [inline package.json:37]: why is this downgraded?

File: package.json [hunk 3]
```diff
@@ -48,7 +49,7 @@
   "scripts": {
     "build": "tsc",
     "format": "prettier --write .",
-    "test": "ava --timeout=2m build/**/*.test.js",
+    "test": "npm run build && ava --timeout=2m build/**/*.test.js",
```
Comment (cdaringe) [inline package.json:52]: please drop this

File: src/__tests__/fixture/db.ts [hunk 4]
```diff
@@ -30,7 +30,6 @@ export async function imageExists(imageName: string) {
 }
 
 export async function purgeContainer(container: Docker.Container) {
-  debugger;
```
Comment (cdaringe) [inline src/__tests__/fixture/db.ts:33]: ❤️  thx. whoopsie!

File: src/__tests__/main.test.ts [hunk 5]
```diff
@@ -32,7 +32,9 @@ test.beforeEach(async (t) => {
       weight real,
       make varchar,
       model varchar,
-      primary key (id)
+      serial_key varchar,
```
Comment (cdaringe) [inline src/__tests__/main.test.ts:35]: a [`serial_number`](https://thebestbikelock.com/bike-serial-number/) is an actual entity with bikes. i have not heard it referred to as a `serial_key`. willing to use `serial_number` or `serial_num`?

File: src/__tests__/main.test.ts [hunk 6]
```diff
@@ -134,23 +136,26 @@ const fetchAllRoles = async (t: PluginExecutionContext) => {
...
 };
 
-const create = async (t: PluginExecutionContext) =>
-  execGqlOp(
-    t,
-    nanographql`
-  mutation {
+const create = async (
+  t: PluginExecutionContext,
+  extraProperties: { [key: string]: any } = {}
```
Comment (cdaringe) [inline src/__tests__/main.test.ts:141]: please use a real type vs `any`

File: package.json (hunk 1 above)
Comment (mgagliardo91) [reply package.json:21]: The previous dependency was behind and we were breaking due to conflicting dependencies in our tree. This one appears to be more regularly up-to-date, though I'm not positive if it changes the output in circle much.

File: package.json (hunk 2 above)
Comment (mgagliardo91) [reply package.json:37]: graphile-build constraint: https://github.com/graphile/graphile-engine/blob/v4/package.json#L49

File: package.json (hunk 3 above)
Comment (mgagliardo91) [reply package.json:52]: Sure - though basing tests on the compiled code can be quite frustrating if the test suite isn't building it as well. I'll revert, but this made it a lot easier to write the tests.

**Comment by mgagliardo91** [comment]
@cdaringe pushed a commit to resolve your feedback, thanks!

**Comment by cdaringe** [comment]
Just writing to let you know, I haven't forgotten about this. It's prompted me to clone your branch, add some extra stuff. Will get it merged in (with your commits) or ask more questions soon. Sorry for the delay

File: src/postgraphile-upsert.ts [hunk 7]
```diff
@@ -223,25 +223,104 @@ export const PgMutationUpsertPlugin: Plugin = (builder) => {
...
+                      }),
+                      {}
+                    );
+
+                  const upsertedColumns: Set<string> = new Set(
+                    attributes.filter((attr) => {
+                      if (
+                        Object.prototype.hasOwnProperty.call(
+                          inputData,
+                          fieldName
```
Comment (cdaringe) [inline src/postgraphile-upsert.ts:244]: will this condition ever be true?

`{  model: "kula deluxe v2" weight: 0.0, serialKey: "1234" }` has fieldName `upsertBike`?

**Comment by cdaringe** [comment]
Merged your commits in in #374. added some more TS & an extra case. some tweaks were made. hopefully they work out for ya

**Comment by mgagliardo91** [comment]
@cdaringe Ah, looking at https://github.com/cdaringe/postgraphile-upsert/pull/374, it looks like the only commit that actually made it to `main` was https://github.com/cdaringe/postgraphile-upsert/commit/bb3455fcedba55a2a6c4ac71784db1b8bb43c287. I'm not sure how this happened, but you can see the history in https://github.com/cdaringe/postgraphile-upsert/commits/main

**Comment by mgagliardo91** [comment]
> ...the 3.x version has the good stuff specifically. the change to commitlint-circle added a bit of a hiccup in getting the latest bits released because i didn't realize a behavior it had in main, but we're set now

@cdaringe thanks for jumping on this! I appreciate it. I will try and test it out this weekend, but `main` looks good to me!

**Pull request closed without merging**